### ISSUE TITLE
Upgrade to v0.4.3 cause issues with db:test:prepare command

### ISSUE THREAD

**Issue by diegodurante** [mixonic/ranked-model#139]
Hi there,
we are using this gem since many time and worked fine.
**But now we are facing errors after updating gem from v0.4.0 to v0.4.3.**
I faced an issue with command:
`RAILS_ENV=test bundle exec rake db:test:prepare` that is failing with this error:

````
Ready to execute commands
rake aborted!
ActiveRecord::NoDatabaseError: FATAL:  database "test" does not exist
/pics/app/models/entry/sample_entry.rb:3:in `<class:SampleEntry>'
/pics/app/models/entry/sample_entry.rb:1:in `<top (required)>'
/pics/spec/factories/entries.rb:110:in `block (2 levels) in <top (required)>'
/pics/spec/factories/entries.rb:2:in `block in <top (required)>'
/pics/spec/factories/entries.rb:1:in `<top (required)>'
/pics/config/environment.rb:5:in `<top (required)>'
/usr/local/bin/bundle:23:in `load'
/usr/local/bin/bundle:23:in `<main>'

Caused by:
PG::ConnectionBad: FATAL:  database "test" does not exist
/pics/app/models/entry/sample_entry.rb:3:in `<class:SampleEntry>'
/pics/app/models/entry/sample_entry.rb:1:in `<top (required)>'
/pics/spec/factories/entries.rb:110:in `block (2 levels) in <top (required)>'
/pics/spec/factories/entries.rb:2:in `block in <top (required)>'
/pics/spec/factories/entries.rb:1:in `<top (required)>'
/pics/config/environment.rb:5:in `<top (required)>'
/usr/local/bin/bundle:23:in `load'
/usr/local/bin/bundle:23:in `<main>'
Tasks: TOP => db:test:prepare => environment
(See full trace by running task with --trace)
````

The line where error refer `/pics/app/models/entry/sample_entry.rb:3:in '<class:SampleEntry>'` has this code:

````
class Entry::SampleEntry < Entry
  include RankedModel
  ranks :index_order

  default_scope { where.not(customer_id: nil).rank(:index_order) }
  before_create -> (entry) { entry.skip_moderation! }
end
````

The strange thing is that also with v0.4.3 everything seems working fine, only when we execute `RAILS_ENV=test bundle exec rake db:test:prepare` we face the issue, both locally and in CI environment.

Note:
- I tried to move configuration to parent class
- I tried to rename database column from `index_order` to `row_order` . Also in another model we use `row_order` and it is working fine. Documentation does not specify if you must use `row_order` naming or you can customize it but we tried and seems working both with `row_order` and `index_order`.

**I tried many different things but I was able to fix this issue only by downgrading to v0.4.0.**

This is the **Gemfile**:
````
source 'https://rubygems.org'
ruby '2.4.5'

git_source(:github) do |repo_name|
  repo_name = "#{repo_name}/#{repo_name}" unless repo_name.include?("/")
  "https://github.com/#{repo_name}.git"
end

gem 'rails',                  '4.2.11'
... # many other gems omitted
gem 'ranked-model',           '0.4.0' # upgrade to v0.4.3 cause issue!

group :doc do
  gem 'sdoc', require: false
end

group :development do
  gem 'bullet',              '5.5.0'
  gem 'slim-rails',          '3.1.2'
end

group :development, :test do
  gem 'rspec-rails',        '~> 3.5'
  gem 'byebug',             '9.0.6'
  gem 'rack_session_access'
  gem 'puma',               '3.10.0'
  gem 'factory_bot_rails',  '4.11.1'
  gem 'faker',              '1.7.3'
end

group :test do
  gem 'capybara',           '2.7.1'
  gem 'database_cleaner',   '1.5.3'
  gem 'launchy',            '2.4.3'
  gem 'selenium-webdriver', '3.141.0'
  gem 'email_spec',         '2.1.0'
  gem 'rspec-sidekiq',      '2.2.0'
  gem 'poltergeist',        '1.18.0'
  gem 'rack-test',          '~> 0.6.2'
  gem 'webmock',            '~> 3.0.1'
  gem 'rspec-retry'
  gem 'geckodriver-helper'
end
````

**Comment by brendon** [comment]
Hi @diegodurante, I think I know what is probably going on. We now do a quick check that the rank column doesn't have a default value and raise an error if so: https://github.com/mixonic/ranked-model/blob/200ebe2e5cf283219022688f733d78c79df2fa71/lib/ranked-model.rb#L51

It sounds like this is being run before your database or table actually exists in this case. With this pointer in mind, can you see if it's something you think we need to fix, or something to do with a unique setup in your test harness? I'd assume normally the test database would exist by the time this initialisation code is run.

Let me know how you get on :)

**Comment by diegodurante** [comment]
Hi @brendon ,
> can you see if it's something you think we need to fix, or something to do with a unique setup in your test harness?

I am not sure, anyway I checked and is not present a default value for that column. That column was created with this migration:
````
def change
     add_column :entries, :index_order, :integer
end
````
and also from `structure.sql` file I can see that doesn't have a default value.

Regarding:
> "I'd assume normally the test database would exist by the time this initialisation code is run."

These are the steps we execute in our CI :
````
docker-compose build
docker-compose up db_migrate
docker-compose up -d sidekiq_cron sidekiq app
docker-compose run -e "RAILS_ENV=test" app bundle exec rake db:test:prepare
docker-compose run -e "RAILS_ENV=test" app bundle exec rspec spec --format documentation --format json --out rspec_report.json
````

So yes, your assumption should be true.
Anyway your code is breaking the db test initialization. That's strange.

Also the error log does not help to debug the issue.

I think this should be fixed somehow, or at least explained in documentation what we should do when we want to "rank" a new model while upgrading the gem.

Many thanks.

**Comment by brendon** [comment]
@diegodurante, of course we should fix it :) The problem is that by the time `column_defaults` is called, Rails hasn't yet connected to the database called `test`, or possibly you don't even have a database called `test` at all? We rely on a connection existing to the database at that point in order to query it for the default value (even if there isn't a default value).

We might have to look at delaying this check until request time but that would be more inconvenient. Could you do some more digging around the timing of the connection to your database?

Also, if possible it'd be good to have a PR with a failing test (or a demo app) to isolate the issue. You'll see our test suite successfully runs so there might be something particular to how your app boots up in test mode that delays the database connection.

**Comment by simeonwillbanks** [comment]
I am seeing the same issue while running `RAILS_ENV=production bundle exec rake assets:precompile` since CI hasn't yet configured the database.  

https://github.com/mixonic/ranked-model/blob/db72d7a65ff207b84cf048c25dc7ea5c39e66a49/lib/ranked-model.rb#L51-L53

https://github.com/mixonic/ranked-model/commit/db72d7a65ff207b84cf048c25dc7ea5c39e66a49#diff-2d690414b76bb6b788905001a9caf848R51

If a column doesn't have a default value, what problem would you experience?  Maybe we solve this problem with documentation instead of enforcing a solution in code?

**Comment by brendon** [comment]
@simeonwillbanks, thanks for raising this issue :) I see now that assuming a database connection might not always work. This check was added in to prevent people raising these issues:

- https://github.com/mixonic/ranked-model/issues/133
- https://github.com/mixonic/ranked-model/issues/136
- https://github.com/mixonic/ranked-model/issues/135

Did you have any ideas on how we could do this in a way that is still disruptive for users who have a default value on the rank column (i.e. a hard error) but gets around the problem of the database not existing?

I might look into verifying a connection exists before attempting to implicitly use it.

**Comment by brendon** [comment]
Everyone, can you check out the latest master and see if this solves your issue? It should just ignore database connection issues now.

**Comment by simeonwillbanks** [comment]
@brendon Thank you for the links.  Sounds like ensuring a default value with code is a good idea ❤️ 

> I might look into verifying a connection exists before attempting to implicitly use it.

I like this idea!  Thanks for trying it out.

I'm looking at `master`, and I don't see any recent commits.  

 - https://github.com/mixonic/ranked-model/commits/master

Did you commit on a branch or maybe forget to push?

**Comment by brendon** [comment]
Sorry, my bad! That button in the GitHub client is too multifunctional, I forgot to give it a second push after it fetched :)

**Comment by simeonwillbanks** [comment]
No worries @brendon ❤️ I provided some feedback on 74cc8bd705e9b2bfbaa6514d2718ee374cd13caf

**Comment by brendon** [comment]
All good :) For the others, please try: 
https://github.com/mixonic/ranked-model/commit/b177d762f002f4b4df1a2a42959377ce572c5e3a though it should be functionally the same.

**Comment by simeonwillbanks** [comment]
@brendon b177d762f002f4b4df1a2a42959377ce572c5e3a fixed the issue, and `RAILS_ENV=production bundle exec rake assets:precompile` works

```
# Gemfile
gem 'ranked-model', git: 'https://github.com/mixonic/ranked-model', ref: 'b177d762f002f4b4df1a2a42959377ce572c5e3a'
```

Thank you!  

Will you be cutting a release?

**Comment by brendon** [comment]
Sure am :) Glad to hear it worked for you. I've released 0.4.4 :)
